### ISSUE TITLE
Rule S2871: Update implementation to raise on non-string-like arrays

### ISSUE THREAD

**Issue by yassin-kammoun-sonarsource** [SonarSource/SonarJS#1696]
We currently raise issues for numeric arrays only, but we can improve the rule and extend its scope to raise on non-string-like arrays. We define a string-like data type as being either:

Primitive type `string`:

```typescript
let strings: string[] = ...
strings.sort(); // OK
```

Types (Type, Class, Interface) overriding the `toString` method:

```typescript
interface I {
  toString(): string;
}
let interfaces: I[] = ...
interfaces.sort(); // OK
```

Intersections involving at least one of the above:
```typescript
let intersections: (string & something)[] = ...
intersections.sort(); // OK
```

**Comment by ilia-kebets-sonarsource** [comment]
We probably shouldn't raise issues on strings (and string-likes) either.

**Comment by yassin-kammoun-sonarsource** [comment]
We can close the ticket since the rule implementation arrays of string, arrays of number, and arrays of bigint. Given the TypeScript compiler API, it's difficult to provide something more sophisticated.
